src/tree_ops.jl

```diff
@@ -12,6 +12,175 @@ function show(io::IO, tree::NNTree{V}) where {V}
     print(io,   "  Reordered: ", tree.reordered)
 end
 
+struct NNTreeNode{T <: NNTree, R}
+    index::Int
+    tree::T
```

Comment:
I have a general "philosophy" that storing something big (a full KD Tree) in something that is conceptually small (a tree node) is often a mistake. 

As you traverse the tree you will create all these nodes that will all contain the same `tree`. What do you think about dropping the `tree` field and instead require a user to provide the tree a an argument to the traverse functions?

Reply:
This is a good question and good rationale. My own experience has been that Julia is very good at optimizing the codes when the types are immutable, so I doubt it is really creating different copies if you use it in a function. 

My argument for the current organization is that node ids are tied to the tree and so this makes it so that you don't have an additional argument hanging out everywhere..., it makes it easy and simple to write codes that do the right thing and get the answer right. But as I said, I hadn't considered your particular perspective here.

Is there a test we could do to resolve if this is an issue? (i.e to convince me that your perspective is correct, or for me to convince you it isn't a problem to store the tree and the compiler really is smart enough?) 

Maybe, vectors of nodes would be bad for including the tree? But we do we ever actually need them? 

Another argument for keeping it linked is that the AbstractTrees interface is 'node' oriented, so you define children, parent, etc. on a node level; which would require keeping the tree as part of the struct.

Reply:
> But... if I store a ref to the tree rather than the full tree structure

I don't fully understand what that means.

Reply:
>  This stores a pointer to the tree information instead of a copy of all the information.

That is not how Julia works. It will not store a copy of all the information because a tree is not an "isbitstype".

Reply:
Regarding storing the kd tree inside the node or not I want to make an analogue to the iteration protocol in Julia. For that you have:

- An `iterate` function taking in an **object** + a **state**, giving back a **value** and a ** new state**

For an array this would be:

- The object is the array
- The state is the index into the array
- The value is the item of the array at that index
- The function used to traverse the array is `iterate` 

I think it would be good to make something similar here.

- The object we have is the tree.
- The state is the "index" of the node
- The value is a "node" 
- Instead of the `iterate` function we have a collection of other function that takes an object + a state and returns a node + a new state.

The current design feels kind of "object oriented" in that you create this one big object that you implement all the functions one but I am not sure that is the best in practice.

Reply:
The issue I have with the iterate analogy is that iterate is designed to execute within a single function context -- and has some nice syntax to hide the complexity and different types of objects -- whereas most of the tree walking functions are designed to execute recursively, where there is no such affordance that I know of.  So you'll have to pass the tree structure to any subfunction -- as well as the node structure. 

The current design is just designed to be easy to use; it's also feasible to adapt to the AbstractTrees.jl interface (although I haven't done that yet...) where they do the same thing with parent/children/etc. functions. 

But it seems like you are still leaning against it enough though there is minimal overhead, is that correct?

Reply:
Just to be precise, the interface you would like is: 

```
children(T::Tree, n::Node) -> (nl::Node, nr::Node)
parent(T::Tree, n::Node) -> (p::Node)
region(n::Node) -> 
leaf_points(T::Tree, n::Node) -> something that iterates over points in the leaf node
etc...
```

where node is something simple like:

```
struct Node{R} 
  index::Int
  region::R
end 
```

Reply:
Just a quick nudge on this question of interface. Would love to get this wrapped up in the next week or so before some obligations for school starts.

Reply:
Okay, since I had a moment, I just implemented the interface above. As a check, we can do non-recursive exploration of the tree using the current children, parent, next/prev sibling structure, see, the e.g. points iterator...

Reply:
Hi, sorry for the slow response here and sorry for being a bit "annoying" with trying to figure out the "best" interface to use.

A reason for this is that this is my first Julia package so it holds a bit of a special place in my heart and I have also worked quite a bit to reduce memory footprint and improve performance.

I can add your package so it is tested as part of the CI here (and you could then at any time also implement whatever tree walking interface you want there and it will not be broken, or at least it can be updated if changes are made here that would be incompatible with it).